fs/cifs/cifsacl.c

```diff
@@ -1670,3 +1679,1082 @@ id_mode_to_cifs_acl(struct inode *inode, const char *path, __u64 *pnmode,
+						flags,
+						ace);
+		if (error) {
+			return error;
+		}
+
+		ace_sz = CIFS_ACE_SIZE(ace->sid.num_subauth);
+		ace->size = cpu_to_le16(ace_sz);
+		out_sz += ace_sz;
+	}
+
+	if (error) {
+		return error;
+	}
+
+	*size = out_sz;
+
+	return 0;
+}
+
+static int
+convert_zfsacl_to_cifsacl(u32 *aclbuf,
+			  u32 acecnt,
+			  struct inode *inode,
+			  struct cifs_acl *pdacl,
+			  u32 dacl_ace_cnt,
+			  u16 *pacl_size_out)
+{
+	u32 i, nsize = sizeof(struct cifs_acl);
+	char *acl_base = (char *)pdacl;
+	u16 size;
+	int error;
+
+	for (i = 0; i < acecnt; i++) {
+		u32 *zfsace = aclbuf + (i * NACE41_LEN);
+		error = convert_zfsace_to_cifs_aces(zfsace, acl_base + nsize, inode, &size);
+		if (error)
+			return error;
+
+		nsize += size;
+	}
+
+	*pacl_size_out = nsize;
+	pdacl->size = cpu_to_le16(nsize);
+	pdacl->revision = cpu_to_le16(ACL_REVISION);
+	pdacl->num_aces = cpu_to_le32(dacl_ace_cnt);
+
+	return 0;
+}
+
+static void
+force_smb3_dacl_info(struct smb3_sd *sd, u32 acl_flag)
+{
+	u16 control = ACL_CONTROL_SR | ACL_CONTROL_DP;
+
+	if (acl_flag & ACL4_PROTECTED) {
+		control |= ACL_CONTROL_PD;
+	}
+
+	/*
+	 * kzalloc call zero-initialized
+	 * sd->Sbz1, which is correct since we are not
+	 * using resource manager
+	 */
+	sd->Revision = 1;
+	sd->Control = cpu_to_le16(control);
+}
+
+/*
+ * This is special handling for either NULL or emtpy ACLs.
+ * Returns 0 if ACL is generated, -EAGAIN if regular parsing
+ * required, and otherwise -errno.
+ */
+static int
+parse_single_ace(u32 *zfsace, struct cifs_ntsd **ppntsd_out, u32 *acllen_out)
+{
+	u32 perms, flags, iflag, who_id, ace_type;
+	bool dacl_is_null = false, dacl_is_empty = false;
+	u32 secdesclen = sizeof(struct cifs_ntsd);
+	struct cifs_ntsd *pnntsd = NULL;
+	struct cifs_acl *dacl = NULL;
+
+	perms = ntohl(*(zfsace + NA_ACCESS_MASK_OFFSET));
+	flags = ntohl(*(zfsace + NA_FLAG_OFFSET));
+	iflag = ntohl(*(zfsace + NA_IFLAG_OFFSET));
+	who_id = ntohl(*(zfsace + NA_WHO_OFFSET));
+	ace_type = ntohl(*(zfsace + NA_TYPE_OFFSET));
+
+	if ((iflag == ACEI4_SPECIAL_WHO) && (who_id == ACE4_SPECIAL_EVERYONE) &&
+	    (perms == ACE4_ALL_PERMS) && (flags == 0) &&
+	    (ace_type == ACE4_ACCESS_ALLOWED_ACE_TYPE)) {
+		dacl_is_null = true;
+	}
+
+	if ((iflag == ACEI4_SPECIAL_WHO) && (who_id == ACE4_SPECIAL_OWNER) &&
+	    (perms == 0) && (flags == 0) &&
+	    (ace_type == ACE4_ACCESS_ALLOWED_ACE_TYPE)) {
+		dacl_is_empty = true;
+		secdesclen += sizeof(struct cifs_acl);
+	}
+
+	if (!dacl_is_null && !dacl_is_empty) {
+		return -EAGAIN;
+	}
+
+	pnntsd = kzalloc(secdesclen, GFP_KERNEL);
+	if (pnntsd == NULL) {
+		return -ENOMEM;
+	}
+
+	force_smb3_dacl_info((struct smb3_sd *)pnntsd, 0);
+
+	*ppntsd_out = pnntsd;
+	*acllen_out = secdesclen;
+
+	if (dacl_is_null) {
+		return 0;
+	}
+
+	pnntsd->dacloffset = cpu_to_le32(sizeof(struct cifs_ntsd));
+	dacl = (struct cifs_acl *)(pnntsd + sizeof(struct cifs_ntsd));
+	dacl->size = cpu_to_le16(sizeof(struct cifs_acl));
+	dacl->revision = cpu_to_le16(ACL_REVISION);
```

Comment:
This is the `dacl_is_empty`, correct?  If so, add comment:
`/* dacl_is_empty */`